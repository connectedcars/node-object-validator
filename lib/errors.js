function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
