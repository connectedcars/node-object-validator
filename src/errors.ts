export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
