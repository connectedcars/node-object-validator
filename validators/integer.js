const { ValidationError } = require('../lib/errors')

function Integer(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, required = true) {
  return {
    type: 'Integer',
    validate: function(key, val) {
      if (val == null) {
        if (required) {
          return new ValidationError(`Field \`${key}\` (${this.type}) is required`, { key, val })
        } else {
          return
        }
      }
      if (typeof val !== 'number' || !Number.isInteger(val)) {
        return new ValidationError(`Field \`${key}\` (${this.type}) must be an integer (received "${val}")`, {
          key,
          val
        })
      }
      if (val < min || val > max) {
        return new ValidationError(
          `Field \`${key}\` (${this.type}) must be between ${min} and ${max} (received "${val}")`,
          { key, val }
        )
      }
    }
  }
}

module.exports = Integer
