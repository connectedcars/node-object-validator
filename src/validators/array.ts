import { ValidatorBase } from '../common'
import { NotArrayError, RequiredError, ValidationErrorContext, WrongLengthError } from '../errors'
import { ValidatorTypes } from '../types'

export function validateArray(
  value: unknown,
  minLength = 0,
  maxLength = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error[] | null {
  if (!Array.isArray(value)) {
    return [new NotArrayError(`Must be an array (received "${value}")`, context)]
  }
  if ((minLength !== 0 && value.length < minLength) || value.length > maxLength) {
    return [
      new WrongLengthError(
        `Must contain between ${minLength} and ${maxLength} entries (found ${value.length})`,
        context
      )
    ]
  }
  return null
}

export class RequiredArray<T extends ValidatorTypes = ValidatorTypes> extends ValidatorBase {
  public schema: T
  private type: 'RequiredArray' = 'RequiredArray'
  private minLength: number
  private maxLength: number

  public constructor(schema: T, minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    super()
    this.schema = schema
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] | null {
    if (value == null) {
      return [new RequiredError(`Is required`, context)]
    }
    return validateArray(value, this.minLength, this.maxLength, context)
  }
}

export class OptionalArray<T extends ValidatorTypes = ValidatorTypes> extends ValidatorBase {
  public schema: T
  private type: 'OptionalArray' = 'OptionalArray'
  private minLength: number
  private maxLength: number

  public constructor(schema: T, minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    super()
    this.schema = schema
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] | null {
    if (value == null) {
      return null
    }
    return validateArray(value, this.minLength, this.maxLength, context)
  }
}

export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength: number,
  maxLength: number,
  required?: false
): OptionalArray<T>
export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength: number,
  maxLength: number,
  required: true
): RequiredArray<T>
export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength = 0,
  maxLength = Number.MAX_SAFE_INTEGER,
  required = false
): OptionalArray<T> | RequiredArray<T> {
  return required
    ? new RequiredArray<T>(schema, minLength, maxLength)
    : new OptionalArray<T>(schema, minLength, maxLength)
}
