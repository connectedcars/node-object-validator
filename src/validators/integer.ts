import { NotIntegerError, OutOfRangeError, RequiredError, ValidationErrorContext } from '../errors'
import { Validator } from './common'

export function validateInteger(
  value: number,
  min = 0,
  max = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error | null {
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    return new NotIntegerError(`Must be an integer (received "${value}")`, context)
  }
  if (value < min || value > max) {
    return new OutOfRangeError(`Must be between ${min} and ${max} (received "${value}")`, context)
  }
  return null
}

export class RequiredInteger implements Validator {
  private type: 'RequiredInteger' = 'RequiredInteger'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    this.min = min
    this.max = max
  }

  public validate(value: number, context?: ValidationErrorContext): Error | null {
    if (value == null) {
      return new RequiredError(`Is required`, context)
    }
    return validateInteger(value, this.min, this.max, context)
  }
}

export class OptionalInteger implements Validator {
  private type: 'OptionalInteger' = 'OptionalInteger'
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER) {
    this.min = min
    this.max = max
  }

  public validate(value: number, context?: ValidationErrorContext): Error | null {
    if (value == null) {
      return null
    }
    return validateInteger(value, this.min, this.max, context)
  }
}

export function Integer(required?: false): OptionalInteger
export function Integer(required: true): RequiredInteger
export function Integer(required = false): OptionalInteger | RequiredInteger {
  return required ? new RequiredInteger() : new OptionalInteger()
}
