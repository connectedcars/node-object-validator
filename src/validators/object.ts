import {
  CodeGenResult,
  ValidateOptions,
  ValidatorBase,
  ValidatorBaseOptions,
  ValidatorExportOptions,
  ValidatorOptions
} from '../common'
import { NotObjectFail, RequiredFail, ValidationFailure } from '../errors'
import { toSnakeCase } from '../util'

export function isObject<T extends ObjectSchema>(
  schema: T,
  value: unknown,
  context?: string
): value is ObjectWrap<UndefinedToOptional<{ [K in keyof T]: T[K]['tsType'] }>> {
  const errors = validateObject(schema, value, context, { earlyFail: true })
  if (errors.length === 0) {
    return true
  }
  return false
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isPlainObject(value: unknown): value is Record<string | number | symbol, any> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

export function validateObject(
  schema: Record<string, ValidatorBase>,
  value: unknown,
  context?: string,
  options?: ValidateOptions
): ValidationFailure[] {
  const errors: ValidationFailure[] = []
  if (!isPlainObject(value)) {
    errors.push(new NotObjectFail(`Must be an object`, value, context))
    return errors
  }
  for (const key of Object.keys(schema)) {
    const validator = schema[key]
    const keyName = context ? `${context}['${key}']` : key
    errors.push(...validator.validate(value[key], keyName, { optimized: false, earlyFail: false, ...options }))
    if (options?.earlyFail && errors.length > 0) {
      return errors
    }
  }
  return errors
}

export type ObjectSchema = Record<string, ValidatorBase>

// https://github.com/Microsoft/TypeScript/issues/26705
export type IsUndefined<T, K> = undefined extends T ? K : never
export type IsNotUndefined<T, K> = undefined extends T ? never : K
export type UndefinedKeys<T> = { [K in keyof T]-?: IsUndefined<T[K], K> }[keyof T]
export type NotUndefinedKeys<T> = { [K in keyof T]-?: IsNotUndefined<T[K], K> }[keyof T]
export type IncludeUndefinedTypes<T extends Record<string, unknown>> = { [K in UndefinedKeys<T>]: T[K] }
export type ExcludeUndefinedTypes<T extends Record<string, unknown>> = { [K in NotUndefinedKeys<T>]: T[K] }
export type UndefinedToOptional<T extends Record<string, unknown>> = ExcludeUndefinedTypes<T> &
  Partial<IncludeUndefinedTypes<T>>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ObjectWrap<T> = T extends Record<string, any> ? { [K in keyof T]: T[K] } : never

export abstract class ObjectValidator<T extends ObjectSchema = never, O = never> extends ValidatorBase<
  ObjectWrap<UndefinedToOptional<{ [K in keyof T]: T[K]['tsType'] }>> | O
> {
  public schema: ObjectSchema
  private typeStringGenerated: boolean

  public constructor(schema: T, options?: ValidatorBaseOptions) {
    super(options)
    this.typeStringGenerated = false
    this.schema = schema
    if (options?.optimize !== false) {
      this.optimize(schema)
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, \`${context}\`` : ', context'
    const objValueRef = `objValue${id()}`
    const schemaRef = `scheme${id()}`
    let imports: { [key: string]: unknown } = {
      NotObjectFail: NotObjectFail,
      RequiredFail: RequiredFail
    }
    const declarations = [`const ${schemaRef} = ${validatorRef}.schema`]

    const propValidationCode: string[] = []
    for (const key of Object.keys(this.schema)) {
      const validator = this.schema[key]
      const propName = context ? `${context}['${key}']` : key
      const [propImports, propDeclarations, propCode] = validator.codeGen(
        `${objValueRef}['${key}']`,
        `${schemaRef}['${key}']`,
        id,
        propName,
        earlyFail
      )
      imports = { ...imports, ...propImports }
      declarations.push(...propDeclarations)
      propValidationCode.push(...propCode)
    }

    // prettier-ignore
    const code = [
      `const ${objValueRef} = ${valueRef}`,
      ...this.nullCheckWrap([
      `  if (typeof ${objValueRef} === 'object' ${!this.nullable ? `&& ${objValueRef} !== null ` : ''}&& !Array.isArray(${objValueRef})) {`,
      ...propValidationCode.map(l => `    ${l}`),
      `  } else {`,
      `    errors.push(new NotObjectFail(\`Must be an object\`, ${objValueRef}${contextStr}))`,
      `  }`,
      ], objValueRef, contextStr),
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]

    return [imports, declarations, code]
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString(options)
    }
  }

  protected validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    return validateObject(this.schema, value, context, { earlyFail: this.earlyFail, ...options })
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        const lines = Object.keys(this.schema).map(k =>
          `'${k}': ${this.schema[k].toString(options)}`.replace(/(^|\n)/g, '$1  ')
        )
        let typeStr = `{\n${lines.join('\n')}\n}`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        if (options?.rustTypeName === undefined) {
          throw new Error(`'rustTypeName' option is not set`)
        }
        if (!this.typeStringGenerated) {
          this.typeStringGenerated = true

          const lines = Object.keys(this.schema).map(k => `  ${toSnakeCase(k)}: ${this.schema[k].toString(options)},`)
          return `struct ${options?.rustTypeName} {\n${lines.join('\n')}\n}`
        }
        const isOption = !this.required || this.nullable
        return isOption ? `Option<${options.rustTypeName}>` : `${options.rustTypeName}`
      }
      default: {
        throw new Error(`Language: '${options?.language}' unknown`)
      }
    }
  }

  private constructorString(options?: ValidatorExportOptions): string {
    const lines = Object.keys(this.schema).map(k =>
      `'${k}': ${this.schema[k].toString(options)}`.replace(/(^|\n)/g, '$1  ')
    )
    const schemaStr = `{\n${lines.join(',\n')}\n}`
    const optionsStr = this.optionsString !== '' ? `, ${this.optionsString}` : ''
    return `new ${this.constructor.name}(${schemaStr}${optionsStr})`
  }
}

export class RequiredObject<T extends ObjectSchema> extends ObjectValidator<T> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options })
  }
}

export class OptionalObject<T extends ObjectSchema> extends ObjectValidator<T, undefined> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, required: false })
  }
}

export class NullableObject<T extends ObjectSchema> extends ObjectValidator<T, null> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, nullable: true })
  }
}

export class OptionalNullableObject<T extends ObjectSchema> extends ObjectValidator<T, undefined | null> {
  public constructor(schema: T, options?: ValidatorOptions) {
    super(schema, { ...options, required: false, nullable: true })
  }
}
