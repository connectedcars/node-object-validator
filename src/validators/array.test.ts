import { RequiredFail } from '../errors'
import { ArrayValidator, OptionalArray, RequiredArray, validateArray } from './array'
import { RequiredInteger } from './integer'
import { RequiredObject } from './object'

describe.each([false, true])('Array (optimize: %s)', optimize => {
  describe('validateArray', () => {
    it('should validate simple array', () => {
      const errors = validateArray(new RequiredInteger(), [1, 2, 3, 4])
      expect(errors).toEqual([])
    })
  })

  describe('ArrayValidator', () => {
    it('should validate and give correct result', () => {
      const arrayValidator = new ArrayValidator(new RequiredInteger(), 0, 10, { optimize })
      const str = arrayValidator.validate.toString()
      if (optimize) {
        expect(str).toMatch(/generatedFunction = true/)
      } else {
        expect(str).not.toMatch(/generatedFunction = true/)
      }
      const errors = arrayValidator.validate([1, 2, 4, 5])
      expect(errors).toEqual([])
    })
  })

  describe('RequiredArray', () => {
    it('accepts empty value', () => {
      // TODO: Support multidimensional [{ .Scheme }, RequiredString(), etc. ]
      const validator = new RequiredArray(new RequiredObject({}), 0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(null)).toStrictEqual([new RequiredFail('Is required')])
      expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required')])
    })
  })

  describe('OptionalArray', () => {
    it('accepts empty value', () => {
      const validator = new OptionalArray(new RequiredObject({}), 0, Number.MAX_SAFE_INTEGER, { optimize })
      expect(validator.validate(null)).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
    })
  })
})
