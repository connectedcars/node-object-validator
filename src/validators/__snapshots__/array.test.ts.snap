// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array ArrayValidator should return an function body 1`] = `
Array [
  Object {
    "NotArrayFail": [Function],
    "NotIntegerFail": [Function],
    "OutOfRangeFail": [Function],
    "RequiredFail": [Function],
    "WrongLengthFail": [Function],
  },
  Array [
    "const scheme4 = validator1.schema",
  ],
  Array [
    "const arrayValue1 = value1",
    "if (arrayValue1 != null) {",
    "  if (Array.isArray(arrayValue1)){",
    "    if (arrayValue1.length < 10) {",
    "      for (const [i2, item3] of arrayValue1.entries()) {",
    "        const value5 = item3",
    "        if (value5 != null) {",
    "          if (typeof value5 === 'number' && Number.isInteger(value5)) {",
    "            if (value5 < -9007199254740991 || value5 > 9007199254740991) {",
    "              errors.push(new OutOfRangeFail(\`Must be between -9007199254740991 and 9007199254740991\`, value5, \`[\${i2}]\`))",
    "            }",
    "          } else {",
    "            errors.push(new NotIntegerFail(\`Must be an integer\`, value5, \`[\${i2}]\`))",
    "          }",
    "        } else {",
    "          errors.push(new RequiredFail(\`Is required\`, value5, \`[\${i2}]\`))",
    "        }",
    "      }",
    "    } else {",
    "      errors.push(new WrongLengthFail(\`Must contain between 0 and 10 entries (found \${arrayValue1.length})\`, arrayValue1, context))",
    "    }",
    "  } else {",
    "    errors.push(new NotArrayFail(\`Must be an array\`, arrayValue1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, arrayValue1, context))",
    "}",
  ],
]
`;
